packages/keys/download-keys.ts

```diff
@@ -0,0 +1,119 @@
+      .on('error', reject)
+      .on('exit', (code, signal) => {
+        if (code) process.exit(code);
+        if (signal) process.kill(process.pid, signal);
+        resolve();
+      });
+  });
+
+const __dirname = path.dirname(fileURLToPath(import.meta.url));
+
+let workspace: string | undefined;
+workspace =
+  (await import('workspace-root').then(
+    ({ workspaceRoot }) => workspaceRoot(),
+    () => null, // package not present, or no root
+  )) || undefined;
+
+const args = [undefined, ...process.argv];
+
+const outputOption = args[args.indexOf('--output') + 1];
+const versionOption = args[args.indexOf('--version') + 1];
+const clobberOption = args.includes('--clobber') || args.includes('--force');
+const rmFailedOption = args.includes('--rm-failed');
+const checksumsOption = args[args.indexOf('--checksums') + 1];
+const retryOption = args.includes('--retry');
+
+const versionTag = versionOption || 'v' + version;
+
+const ghRepoRoot = `https://github.com/penumbra-zone/penumbra/raw/${versionTag}`;
+const repoPath = 'crates/crypto/proof-params/src/gen';
+const remoteSource = path.join(ghRepoRoot, repoPath);
+
+const outDir = path.resolve(outputOption ?? path.join(workspace ?? __dirname, 'keys', versionTag));
+const checksumsFile = path.resolve(
+  checksumsOption ?? path.join(__dirname, `${versionTag}.sha256sum`),
+);
+
+const checksums = (await fs.readFile(checksumsFile))
+  .toString()
+  .split('\n')
+  .filter(id => id) // remove blank line
+  .map((line: string) => {
+    const [_line, sha256sum, name] = line.match(/^([0-9a-f]{64}) {2}([a-z_]+_pk.bin)$/)!;
+    return [sha256sum, name] as [string, string];
+  });
+
+const fileNames = checksums.map(([_, name]) => name);
+
+let retry = Number(retryOption);
+
+async function downloadKeys() {
+  // download
+  await cmd('curl', [
+    '--parallel',
+    '--create-dirs',
+    ...['--output-dir', outDir],
+    ...(clobberOption
+      ? // clobber will overwrite existing files
+        ['--clobber']
+      : // continue-at will resume or skip existing files
+        ['--continue-at', '-']),
+    '--location',
+    '--remote-name',
+    // curl supports url globbing
+    `${remoteSource}/{${fileNames.join(',')}}`,
+  ]);
+}
+
+function check() {
+  return checksums.map(async ([expectSha, name]) => {
+    const filePath = path.resolve(outDir, name);
+    const checkSha = createHash('sha256')
+      .update(await fs.readFile(filePath))
+      .digest('hex');
+
+    if (expectSha !== checkSha) {
+      console.log(`${name}: FAILED`);
+      process.exitCode ??= 0;
+      process.exitCode++;
+      if (rmFailedOption || (retryOption && !clobberOption)) fs.rm(filePath);
+      throw new Error(`${name}: ${checkSha}`);
+    } else console.log(`${name}: OK`);
+  });
+}
+
+do {
+  await downloadKeys();
+  const checks = check();
+  await Promise.allSettled(checks);
+  await Promise.all(checks).then(
+    () => {
+      retry = 0;
+      stdout.log(outDir);
```

Comment:
the only thing this entire script echoes to stdout path to the dir containing a successful key download. if it fails, nothing to stdout. all other output is status on stderr